File: RemoveInvalidParentheses3.cpp
#include<iostream>
#include<vector>
#include<unordered_set>
#include<queue>

using namespace std;

class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
    	vector<string> result;
    	helper(s, 0, 0, {'(', ')'}, result);
    	return result;
    }
    
    void helper(string s, int last_i, int last_j, vector<char> p, vector<string>& result) {
    	int cnt = 0;
    	for(int i = last_i;i < s.size();i++) {
    		if(s[i] == p[0]) cnt++;
    		else if(s[i] == p[1]) cnt--;
    		if(cnt >= 0) continue;
    		for(int j = last_j;j <= i;j++) {
    			if(s[j] == p[1] && (j == last_j || s[j] != s[j - 1])) {
    				helper(s.substr(0, j) + s.substr(j + 1), i, j, p, result);
				}
			}
			return;
		}
		
		string rev = string(s.rbegin(), s.rend());
		if(p[0] == '(') helper(rev, 0, 0, {')', '('}, result);
		else result.push_back(rev);
	}
    
};

int main() {
	string s;
	cin>>s;
	
	Solution *solution = new Solution();
	vector<string> result = solution->removeInvalidParentheses(s);
	
	for(int i=0;i<result.size();i++) cout<<result[i]<<endl;
	
	return 0;
}

File: RemoveInvalidParentheses3.py
class Solution:
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        def isValid(s):
            cnt = 0
            for ch in s:
                if ch == '(':
                    cnt += 1
                elif ch == ')':
                    cnt -= 1
                    if cnt < 0:
                        return False
            return cnt == 0

        def helper(s, start, cnt1, cnt2):
            if cnt1 == 0 and cnt2 == 0:
                if isValid(s):
                    result.append(s)
                return
            for i in range(start, len(s)):
                if i != start and s[i] == s[i - 1]:
                    continue
                else:
                    if cnt1 > 0 and s[i] == '(':
                        helper(s[: i] + s[i + 1:], i, cnt1 - 1, cnt2)
                    elif cnt2 > 0 and s[i] == ')':
                        helper(s[: i] + s[i + 1:], i, cnt1, cnt2 - 1)

        result = []
        cnt1 = 0
        cnt2 = 0
        for c in s:
            if c == '(':
                cnt1 += 1
            elif c == ')':
                if cnt1 == 0:
                    cnt2 += 1
                else:
                    cnt1 -= 1
        helper(s, 0, cnt1, cnt2)

        return result


if __name__ == '__main__':
    s = input()
    solution = Solution()
    print(solution.removeInvalidParentheses(s))
